[SYSTEMML-963] Updated license for standalone-jar

gweidner opened the pull request:
Added BSD license text for abego.treelayout and stringtemplate.v4 since classes from both are contained in standalone-jar from ANTLR.

gweidner commented:
The abego license text was restored from previous version and the stringtemplate license text based on http://www.stringtemplate.org/license.html which matches the ANTRL license at http://www.antlr.org/license.html except for names listed after copyright year.

cc @deroneriksson

deroneriksson commented:
Great catch @gweidner! Looks like for the ANTLR 4.5.3 that rather than using transitive dependencies, they actually pack org.abego.* and org.stringtemplate.* into the actual antlr4-4.5.3.jar file. So you are correct that these need to be added to the license for the standalone jar.

LGTM.

I think this will also need to be added to some of our other artifacts (the license for antlr which I recently removed and the additional license for stringtemplate). I will investigate and go ahead and do this.

gweidner commented:
Thanks for the review @deroneriksson! Before updating licenses (including the one in this PR), let me check if the ANTRL4 Tool jar can be removed from pom for standalone which would help reduce license dependencies.

deroneriksson commented:
I would be surprised if it can be removed, but definitely check.

deroneriksson commented:
Actually, I see lots of "import org.antlr.v4.runtime.*" imports in the project, so that might actually work (since the runtime is a separate dependency).

deroneriksson commented:
I will also need to check for https://github.com/apache/incubator-systemml/pull/263. Please let me know how your test goes.

gweidner commented:
As an alternative approach, I changed the standalone-jar.xml such that the ANTLR Tool jar (antlr4-4.5.3.jar) classes were no longer bundled.  Extracting the contents of the resulting standalone-jar no longer included org.abego or org.stringtemplate classes.  So I also removed the StringTemplate license, abego TreeLayout license, and reference to ANTLR 4 Tool from standalone-jar's license text.

Sanity tested the standalone-jar on Windows from command line:

C:\git\incubator-systemml\target>java -jar systemml-0.11.0-incubating-SNAPSHOT-standalone.jar -s "print('hello world');"
16/09/28 15:51:45 INFO api.DMLScript: BEGIN DML run 09/28/2016 15:51:45
16/09/28 15:51:45 INFO api.DMLScript: HADOOP_HOME: C:\git\incubator-systemml\target\lib\hadoop
16/09/28 15:51:45 INFO conf.DMLConfig: Using internal default configuration settings.  If you wish to customize any settings, please supply a `SystemML-config.x
ml` file.
16/09/28 15:51:45 WARN hops.OptimizerUtils: Auto-disable multi-threaded text read for 'text' and 'csv' due to thread contention on JRE < 1.8 (java.version=1.7.0
).
hello world
16/09/28 15:51:46 INFO api.DMLScript: SystemML Statistics:
Total execution time:           0.783 sec.
Number of executed MR Jobs:     0.

16/09/28 15:51:46 INFO api.DMLScript: END DML run 09/28/2016 15:51:46

gweidner commented:
Test build in progress at [PR build 707](https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/707/)

deroneriksson commented:
@gweidner I also investigated this and saw similar results (that we can remove the antlr tool jar). I will also update my JIRA/PR for the standalone tar.gz and zip, which have a similar situation.

deroneriksson commented:
Assuming the tests pass, LGTM.

The contents of the the standalone jar look good (the only antlr classes I see are the runtime classes). The LICENSE also has been updated to not include the antlr tool library. I also tested the standalone jar with a 'hello world' example and a univariate stats example and both worked successfully.

gweidner commented:
Thank you @deroneriksson!
Also tested Univar-stats.dml example from your https://github.com/apache/incubator-systemml/pull/264 with results pasted below.

C:\git\incubator-systemml\target>java -jar systemml-0.11.0-incubating-SNAPSHOT-standalone.jar -f ../scripts/algorithms/Univar-Stats.dml -nvargs X=data/haberman.
data TYPES=data/types.csv STATS=data/univarOut.mtx CONSOLE_OUTPUT=TRUE
16/09/28 17:50:23 INFO api.DMLScript: BEGIN DML run 09/28/2016 17:50:23
16/09/28 17:50:23 INFO api.DMLScript: HADOOP_HOME: C:\git\incubator-systemml\target\lib\hadoop
16/09/28 17:50:23 INFO conf.DMLConfig: Using internal default configuration settings.  If you wish to customize any settings, please supply a `SystemML-config.x
ml` file.
16/09/28 17:50:23 WARN hops.OptimizerUtils: Auto-disable multi-threaded text read for 'text' and 'csv' due to thread contention on JRE < 1.8 (java.version=1.7.0
).
-------------------------------------------------
Feature [1]: Scale
 (01) Minimum             | 30.0
 (02) Maximum             | 83.0
 (03) Range               | 53.0
 (04) Mean                | 52.45751633986928
 (05) Variance            | 116.71458266366658
 (06) Std deviation       | 10.803452349303281
 (07) Std err of mean     | 0.6175922641866753
 (08) Coeff of variation  | 0.20594669940735139
 (09) Skewness            | 0.1450718616532357
 (10) Kurtosis            | -0.6150152487211726
 (11) Std err of skewness | 0.13934809593495995
 (12) Std err of kurtosis | 0.277810485320835
 (13) Median              | 52.0
 (14) Interquartile mean  | 52.16013071895425
-------------------------------------------------
Feature [2]: Scale
 (01) Minimum             | 58.0
 (02) Maximum             | 69.0
 (03) Range               | 11.0
 (04) Mean                | 62.85294117647059
 (05) Variance            | 10.558630665380907
 (06) Std deviation       | 3.2494046632238507
 (07) Std err of mean     | 0.18575610076612029
 (08) Coeff of variation  | 0.051698529971741194
 (09) Skewness            | 0.07798443581479181
 (10) Kurtosis            | -1.1324380182967442
 (11) Std err of skewness | 0.13934809593495995
 (12) Std err of kurtosis | 0.277810485320835
 (13) Median              | 63.0
 (14) Interquartile mean  | 62.80392156862745
-------------------------------------------------
Feature [3]: Scale
 (01) Minimum             | 0.0
 (02) Maximum             | 52.0
 (03) Range               | 52.0
 (04) Mean                | 4.026143790849673
 (05) Variance            | 51.691117539912135
 (06) Std deviation       | 7.189653506248555
 (07) Std err of mean     | 0.41100513466216837
 (08) Coeff of variation  | 1.7857418611299172
 (09) Skewness            | 2.9546334710883215
 (10) Kurtosis            | 11.425776549251449
 (11) Std err of skewness | 0.13934809593495995
 (12) Std err of kurtosis | 0.277810485320835
 (13) Median              | 1.0
 (14) Interquartile mean  | 1.2483660130718954
-------------------------------------------------
Feature [4]: Categorical (Nominal)
 (15) Num of categories   | 2
 (16) Mode                | 1
 (17) Num of modes        | 1
16/09/28 17:50:25 INFO api.DMLScript: SystemML Statistics:
Total execution time:           0.579 sec.
Number of executed MR Jobs:     0.

16/09/28 17:50:25 INFO api.DMLScript: END DML run 09/28/2016 17:50:25

deroneriksson commented:
BTW, the antlr4-4.5.3.jar is 1.5M in size, so removing this unnecessary dependency sizably shrinks the standalone jar (33.3M to 32.2M).

gweidner commented:
Good point Deron re: jar size.
